Mathlib/GroupTheory/CommutingProbability.lean

```diff
@@ -132,3 +136,86 @@ theorem inv_card_commutator_le_commProb : (↑(Nat.card (commutator G)))⁻¹ 
+  congr
+
+private lemma div_two_lt {n : ℕ} (h0 : n ≠ 0) : n / 2 < n :=
+  Nat.div_lt_self (Nat.pos_of_ne_zero h0) (lt_add_one 1)
+
+private lemma div_four_lt : {n : ℕ} → (h0 : n ≠ 0) → (h1 : n ≠ 1) → n / 4 + 1 < n
+  | 0 | 1 | 2 | 3 => by decide
+  | n + 4 => by intros; linarith [n.add_div_right four_pos, n.div_le_self 4]
+
+/-- A list of Dihedral groups whose product will have commuting probability `1 / n`. -/
+def reciprocalFactors (n : ℕ) : List ℕ :=
+  if h0 : n = 0 then [0]
+  else if h1 : n = 1 then []
+  else if 2 ∣ n then
+    have := div_two_lt h0
+    3 :: reciprocalFactors (n / 2)
+  else
+    have := div_four_lt h0 h1
+    n % 4 * n :: reciprocalFactors (n / 4 + 1)
+
+@[simp] lemma reciprocalFactors_zero : reciprocalFactors 0 = [0] := rfl
+
+@[simp] lemma reciprocalFactors_one : reciprocalFactors 1 = [] := rfl
+
+lemma reciprocalFactors_def :
+    reciprocalFactors n =
+      if h0 : n = 0 then [0]
+      else if h1 : n = 1 then []
+      else if 2 ∣ n then
+        3 :: reciprocalFactors (n / 2)
+      else
+        n % 4 * n :: reciprocalFactors (n / 4 + 1) := by
+  conv_lhs => unfold reciprocalFactors
+
+/-- A finite product of Dihedral groups. -/
+abbrev Product (l : List ℕ) : Type :=
+  ∀ i : Fin l.length, DihedralGroup l[i]
+
+lemma commProb_nil : commProb (Product []) = 1 := by
+  simp [Product, commProb_pi]
+
+lemma commProb_cons (n : ℕ) (l : List ℕ) :
+    commProb (Product (n :: l)) = commProb (DihedralGroup n) * commProb (Product l) := by
+  simp [Product, commProb_pi, Fin.prod_univ_succ]
+
+/-- Construction of a group with commuting probability `1 / n`. -/
+theorem commProb_reciprocal (n : ℕ) :
+    commProb (Product (reciprocalFactors n)) = 1 / n := by
+  rw [reciprocalFactors_def]
+  by_cases h0 : n = 0
+  · rw [dif_pos h0, commProb_cons, commProb_nil, mul_one, h0, Nat.cast_zero, div_zero]
+    apply commProb_eq_zero_of_infinite
+  · by_cases h1 : n = 1
+    · rw [dif_neg h0, dif_pos h1, commProb_nil, h1, Nat.cast_one, div_one]
+    · by_cases h2 : 2 ∣ n
+      · have := div_two_lt h0
+        rw [dif_neg h0, dif_neg h1, if_pos h2, commProb_cons, commProb_reciprocal (n / 2),
+            commProb_odd (by norm_num)]
+        field_simp [h0]
+        norm_num
+      · have := div_four_lt h0 h1
+        rw [dif_neg h0, dif_neg h1, if_neg h2, commProb_cons, commProb_reciprocal (n / 4 + 1)]
+        have key : n % 4 = 1 ∨ n % 4 = 3 := Nat.odd_mod_four_iff.mp (Nat.two_dvd_ne_zero.mp h2)
+        have hn : Odd (n % 4) := by rcases key with h | h <;> rw [h] <;> norm_num
+        rw [commProb_odd (hn.mul (Nat.odd_iff.mpr (Nat.two_dvd_ne_zero.mp h2)))]
+        rw [div_mul_div_comm, mul_one, div_eq_div_iff, one_mul] <;> norm_cast
+        · have h0 : (n % 4) ^ 2 + 3 = n % 4 * 4 := by rcases key with h | h <;> rw [h] <;> norm_num
+          have h1 := (Nat.div_add_mod n 4).symm
+          zify at h0 h1 ⊢
+          linear_combination (h0 + h1 * (n % 4)) * n
+        · have := hn.pos.ne'
+          positivity
```

Comment:
I think it's totally fine (or even preferable) to unindent chained `by_cases` here to match the `else if`s:
```suggestion
  by_cases h0 : n = 0
  · rw [dif_pos h0, commProb_cons, commProb_nil, mul_one, h0, Nat.cast_zero, div_zero]
    apply commProb_eq_zero_of_infinite
  by_cases h1 : n = 1
  · rw [dif_neg h0, dif_pos h1, commProb_nil, h1, Nat.cast_one, div_one]
  by_cases h2 : 2 ∣ n
  · have := div_two_lt h0
    rw [dif_neg h0, dif_neg h1, if_pos h2, commProb_cons, commProb_reciprocal (n / 2),
        commProb_odd (by norm_num)]
    field_simp [h0]
    norm_num
  · have := div_four_lt h0 h1
    rw [dif_neg h0, dif_neg h1, if_neg h2, commProb_cons, commProb_reciprocal (n / 4 + 1)]
    have key : n % 4 = 1 ∨ n % 4 = 3 := Nat.odd_mod_four_iff.mp (Nat.two_dvd_ne_zero.mp h2)
    have hn : Odd (n % 4) := by rcases key with h | h <;> rw [h] <;> norm_num
    rw [commProb_odd (hn.mul (Nat.odd_iff.mpr (Nat.two_dvd_ne_zero.mp h2)))]
    rw [div_mul_div_comm, mul_one, div_eq_div_iff, one_mul] <;> norm_cast
    · have h0 : (n % 4) ^ 2 + 3 = n % 4 * 4 := by rcases key with h | h <;> rw [h] <;> norm_num
      have h1 := (Nat.div_add_mod n 4).symm
      zify at h0 h1 ⊢
      linear_combination (h0 + h1 * (n % 4)) * n
    · have := hn.pos.ne'
      positivity
```